bumble/host.py

```diff
@@ -152,6 +152,7 @@ class Host(AbortableEventEmitter):
     acl_packet_queue: Optional[AclPacketQueue] = None
     le_acl_packet_queue: Optional[AclPacketQueue] = None
     hci_sink: Optional[TransportSink] = None
+    hci_source: Optional[TransportSource] = None
```

Comment:
This isn't needed. The host doesn't need to keep a reference to the source that's feeding it. It does need to keep a reference to the sink that it feeds, but the source only comes into the picture when calling `set_packet_source`, which is a convenience method for telling the source that its sink is the host object.

Reply:
I used that in the RTK driver when we get invalid packet data, as I noticed sometimes I got a large packet that had a payload of `254`, but never got the remaining packet data for that packet:
```
            if host.hci_source and hasattr(host.hci_source, "parser"):
                host.hci_source.parser.reset()
```
Is there another way we can reset the parser from the driver? The packet I was getting from the dongle had the string `RTK_5` or something like that.

Reply:
Maybe a `parser` property or some other `reset` exposed from the `host()`?

Reply:
@barbibulle, any thoughts on a way of working around this?

Reply:
Sorry for the delay. There are a couple issues to work through here:
1/ The source/sink relationship is a push model, and source is a purely abstract concept of "some object that pushes data to a sink". A sink is just an object that has a `on_packet` method. This means that an object like `Host`, which receives HCI data from a "source" doesn't necessarily know what the source is, it just receives packets. The `Host.set_packet_source` method is purely a convenience function that tells the source where to push data (and looks up a `metadata` property on the source, but that isn't a requirement. In theory, one could attach a source to a `Host` without calling that method. This means that looking for an `host.hci_source` may or may not work depending on how objects were instantiated.

2/ Not all sources have a `parser` property. It is common to see it, because most sources are `Transport` implementations, which often need to extract HCI packets from a stream, and they usually use a `PacketParser` for that. But you could very well have a source that doesn't need to parse an HCI stream (for example, a transport where HCI packets arrive already delimited). So calling something like `host.hci_source.parser.reset()` only works for a transport that happens to use a `PacketParser` and happens to store that parser in a property named `parser`. That could break at any time, since that's not a requirement of transport implementations.

But in this case, the problem we need to solve is narrow, so we should be able to address it differently.
I'll try a few things out over the next week to see if I can think of a maintainable way to achieve what's needed here (i.e without having to put too many constraints on what types of objects and classes can be used as HCI data sources).

Reply:
I could break out the changes into two separate PRs:

1. One for the RTK / HCI Source
2. The other for the parser fixes.

Would that be better?

Reply:
Sorry but took so long to follow up. Yes, breaking up in two PRs would help. For the parser changes, I’ll try to find some time over the weekend to come up with an alternative.

Reply:
Update: I though about a maintainable way to allow signaling of errors back to an HCI source so that it can reset itself, without creating an explicit reference from a sink to a source (i.e no storing of `hci_source` in the `Host` object): a source can listen for events from its sink (first testing if the sink is an EventEmitter, since not all are required to be), then, in the RTK driver case here, the driver could have the host emit an error (upon encountering the timeout), which the usb source would get called back for, and reset itself.  But... there's probably an even simpler way: it turns out that the usb source really doesn't need to use an HCI parser, since packets coming from USB are always delimited. The current implementation subclasses ParserSource by convenience, and uses `self.parser.feed_data(packet)` to emit packets to its sink, but that's really not needed. I'll push a PR where the USB source doesn't use a parser anymore, which should solve the issue here, since even after receiving corrupted data in one USB transfer, the next USB transfer should be a whole packet again.

Reply:
Created #534 . Thank you!